02-peer-protocol.md

```diff
@@ -50,10 +67,451 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+malicious attempts. A malicious node attempting to probe a `contributor`'s
+utxo shot, in the best case, is only granted one attempt per node. We limit
+probe attempts across the network by gossiping bad actor's H2 commitments.
+
+PoDLEs will not stop a determined attacker from repeatedly probing a node,
+but it does increase the cost of doing so and effectively rate-limits
+such attempts.
+
+
+### The `tx_add_output` Message
+
+This message contains information for outputs to a funding transaction.
+
+1. type: 67 (`tx_add_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`serial_id`]
+    * [`u64`:`sats`]
+    * [`u16`:`scriptlen`]
+    * [`scriptlen*byte`:`script`]
+
+#### Requirements
+
+Either node:
+  - MAY omit this message
+
+The sending node:
+  - MUST add all sent outputs to the collaborative transaction
+  - MUST ensure the `script` is a standard script
+  - MUST use a unique `serial_id` for each output
+  - MUST specify a `sats` value greater than the dust limit
+
+The receiving node:
+  - MUST add all received outputs to the collaborative transaction
+  - MUST fail the transaction collaboration if:
+    - it receives a duplicate `serial_id`
+    - it receives a `serial_id` from the peer with the incorrect parity
+
+
+#### Rationale
+Each node must have a complete set of the transaction outputs
+to derive the funding transaction and subsequent commitment signatures.
+
+`serial_id` is a randomly chosen number which uniquely identifies this output.
+Outputs in the final transaction will be sorted by `serial_id`.
+
+`sats` is the satoshi value of the output.
+
+`output_info`.`script` is the scriptPubKey for the output.
+
+
+### The `tx_remove_input` Message
+This message removes an input from the transaction
+
+1. type: 68 (`tx_remove_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`serial_id`]
+
+#### Requirements
+
+The sending node:
+  - MUST NOT send a `tx_remove_input` for an input which is not theirs
+
+The receiving node:
+  - MUST remove the indicated input from the final transaction
+  - MAY ignore the message if:
+    - the `serial_id` is novel (has not been communicated previously)
+  - MUST fail the channel if:
+    - the `serial_id` matches one of the receiving node's inputs
+
+
+### The `tx_remove_output` Message
+This message removes an output from the transaction
+
+1. type: 69 (`tx_remove_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`serial_id`]
+
+#### Requirements
+
+The sending node:
+  - MUST NOT send a `tx_remove_output` for an input which is not theirs
+
+The receiving node:
+  - MUST remove the indicated output from the final transaction
+  - MAY ignore the message if:
+    - the `serial_id` is novel (has not been communicated previously)
+  - MUST fail the channel if:
+    - the `serial_id` matches one of the receiving node's outputs
+
+
+### The `tx_complete` Message
+
+This message signals the conclusion of a peer's transaction
+contributions.
```

Comment:
In practice (e.g. the `initiator` and `contributor` contributions diagram above) this seems more like a `tx_ack` than a message signaling "the conclusion of a peer's transaction contributions"

Reply:
`complete` here means "i have completed my currently known set of updates for this transaction"